pyanaconda/modules/payloads/payload/flatpak/flatpak_manager.py

```diff
@@ -0,0 +1,280 @@
+#
+# Root object for handling Flatpak preinstallation
+#
+# Copyright (C) 2024 Red Hat, Inc.
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# the GNU General Public License v.2, or (at your option) any later version.
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY expressed or implied, including the implied warranties of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+
+
+import os
+from typing import List, Optional
+
+import gi
+
+from pyanaconda.anaconda_loggers import get_module_logger
+from pyanaconda.core.configuration.anaconda import conf
+from pyanaconda.core.glib import GError
+from pyanaconda.core.i18n import _
+from pyanaconda.modules.common.errors.installation import PayloadInstallationError
+from pyanaconda.modules.common.task.progress import ProgressReporter
+from pyanaconda.modules.payloads.constants import SourceType
+from pyanaconda.modules.payloads.payload.flatpak.source import (
+    FlatpakRegistrySource,
+    FlatpakStaticSource,
+    NoSourceError,
+)
+from pyanaconda.modules.payloads.source.source_base import PayloadSourceBase, RepositorySourceMixin
+
+gi.require_version("Flatpak", "1.0")
+gi.require_version("Gio", "2.0")
+
+from gi.repository.Flatpak import Installation, Transaction, TransactionOperationType
+
+# We need Flatpak to read configuration files from the target and write
+# to the target system installation. Since we use the Flatpak API
+# in process, we need to do this by modifying the environment before
+# we start any threads. Setting these variables will be harmless if
+# we aren't actually using Flatpak.
+
+# pylint: disable=environment-modify
+os.environ["FLATPAK_DOWNLOAD_TMPDIR"] = os.path.join(conf.target.system_root, "var/tmp")
+# pylint: disable=environment-modify
+os.environ["FLATPAK_CONFIG_DIR"] = os.path.join(conf.target.system_root, "etc/flatpak")
+# pylint: disable=environment-modify
+os.environ["FLATPAK_OS_CONFIG_DIR"] = os.path.join(conf.target.system_root, "usr/share/flatpak")
+# pylint: disable=environment-modify
+os.environ["FLATPAK_SYSTEM_DIR"] = os.path.join(conf.target.system_root, "var/lib/flatpak")
```

Comment:
Are these safe to be used on a running system?

We support a lot of variety environments and one of it is Live or image installation. It would be bad if we break the user system after installation because we have changed the environment variables of the system.

Reply:
These environment variables changes should only affect anaconda and it's child processes, right?

Reply:
> Last I checked setting `os.environ` in a multithreaded process was still generally unsafe, and Anaconda is definitely one. xref https://sourceware.org/bugzilla/show_bug.cgi?id=15607

The import happens before threads are started. I originally had it very explicitly happening at process initialization: https://github.com/owtaylor/anaconda/blob/flatpak-preinstallation-try-1/pyanaconda/modules/payloads/__main__.py but @jkonecny12 didn't like Flatpak code leaking into a generic place.

Reply:
I see your point. In that case you can move it to `payloads/flatpak/flatpak.py`. My point was mainly to move it out of the shared payloads code which is imported to every payload.